src/codegen/compile.ml

```diff
@@ -2711,30 +2711,25 @@ module Object = struct
   let idx_hash_raw env low_bound =
     let name = Printf.sprintf "obj_idx<%d>" low_bound  in
     Func.share_code2 env name (("x", I32Type), ("hash", I32Type)) [I32Type] (fun env get_x get_hash ->
+      let set_x = G.i (LocalSet (nr 0l)) in
       let (set_h_ptr, get_h_ptr) = new_local env "h_ptr" in
 
-      get_x ^^ Heap.load_field hash_ptr_field ^^ set_h_ptr ^^
+      get_x ^^ Heap.load_field hash_ptr_field ^^
 
-      get_x ^^ Heap.load_field size_field ^^
       (* Linearly scan through the fields (binary search can come later) *)
-      from_m_to_n env (Int32.of_int low_bound) (fun get_i ->
-        get_i ^^
-        compile_mul_const Heap.word_size ^^
-        get_h_ptr ^^
-        G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
-        Heap.load_field 0l ^^
-        get_hash ^^
-        G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
-        G.if_ []
-          ( get_i ^^
-            compile_add_const header_size ^^
-            compile_mul_const Heap.word_size ^^
-            get_x ^^
-            G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
-            G.i Return
-          ) G.nop
-      ) ^^
-      E.trap_with env "internal error: object field not found"
+      (* unskew h_ptr and advance both to low bound *)
+      compile_add_const Int32.(of_int (1 + low_bound * to_int Heap.word_size)) ^^ set_h_ptr ^^
+      get_x ^^ compile_add_const Int32.(of_int ((3 + low_bound) * to_int Heap.word_size)) ^^ set_x ^^
+      G.loop_ [] (
+          get_h_ptr ^^ load_unskewed_ptr ^^
+          get_hash ^^ G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
+          G.if_ []
+            (get_x ^^ G.i Return)
+            (get_h_ptr ^^ compile_add_const Heap.word_size ^^ set_h_ptr ^^
+             get_x ^^ compile_add_const Heap.word_size ^^ set_x ^^
```

Comment:
Is there a reason to advance the both ptrs in lock step rather than just do one subtract and add at the end? I guess you are hoping that the fast path is to not iterate at all, and this is then better?
Anyway, this is nice an clear, just asking.

Reply:
(I'd change the title of the PR to reflect what it's now doing)

Reply:
Yes, the reason is that `diff`-`add` is penalising the _fast path_ when the speculation hits. My gut feeling says also that there are less data dependencies this way. But I'll try to code up the other solution too, and benchmark.